File: discipline_client_ts/discipline_android_primitive_ts/rules/countdown_rule.ts
import { Countdown, Branded } from "../x.ts";

const BRAND = Symbol();

export type CountdownRule = Branded<typeof BRAND, {
  readonly countdown: Countdown,
  readonly isEnabled: boolean,
}>;

export const construct = (isEnabled: boolean, countdown: Countdown): CountdownRule => {
  return Branded(BRAND, {
    countdown,
    isEnabled,
  });
};

export const create = (countdown: Countdown): CountdownRule => {
  return construct(false, countdown);
};

export const getCountdown = (it: CountdownRule): Countdown => {
  return it.countdown;
};

export const CountdownRule = {
  construct,
  create,
  getCountdown,
};

File: discipline4/app/src/main/java/com/example/app/rules/CountdownRule.kt
package com.example.app

import com.example.app.*
import androidx.room.Entity

/**
 * A rule based on a countdown timer
 */
@Entity
public data class CountdownRule private constructor(
  var isEnabled: Boolean,
  val countdown: Countdown
) {
  companion object {
    fun create(countdown: Countdown): CountdownRule {
      return CountdownRule(
        isEnabled = false, 
        countdown = countdown,
      )
    }

    fun construct(
      isEnabled: Boolean, 
      countdown: Countdown,
    ): CountdownRule {
      return CountdownRule(isEnabled, countdown)
    }
  }
  
  fun getIsEnabled(): Boolean {
    return isEnabled
  }

  fun setIsEnabled(newValue: Boolean) {
    isEnabled = newValue
  }

  fun getCountdown(): Countdown {
    return countdown
  }

  /**
   * Checks if this rule is active (enabled and countdown not finished)
   */
  fun isActive(now: Instant): Boolean {
    return isEnabled && !countdown.isFinished(now)
  }
  
  override fun toString(): String {
    return "CountdownRule(enabled=$isEnabled, countdown=${countdown.getTotalDuration()})"
  }
}